src/object/srv_ec_aggregate.c

```diff
@@ -0,0 +1,1207 @@
+ */
+static int
+agg_fetch_data_stripe(struct ec_agg_entry *entry)
+{
+	daos_iod_t	iod = { 0 };
+	daos_recx_t	recx = { 0 };
+	unsigned int	len = entry->ae_oca->u.ec.e_len;
+	unsigned int	k = entry->ae_oca->u.ec.e_k;
+	int		rc = 0;
+
+	rc = agg_prep_sgl(entry);
+	if (rc)
+		goto out;
+	recx.rx_idx = entry->ae_cur_stripe.as_stripenum * k * len;
+	recx.rx_nr = k * len;
+	iod.iod_name = entry->ae_akey;
+	iod.iod_type = DAOS_IOD_ARRAY;
+	iod.iod_size = entry->ae_rsize;
+	iod.iod_nr = 1;
+	iod.iod_recxs = &recx;
+	entry->ae_sgl->sg_nr = 1;
+	entry->ae_sgl->sg_iovs[AGG_IOV_DATA].iov_len =
+						len * k * entry->ae_rsize;
+
+	rc = vos_obj_fetch(entry->ae_chdl, entry->ae_oid,
+			   entry->ae_cur_stripe.as_hi_epoch,
+			   VOS_FETCH_RECX_LIST, &entry->ae_dkey, 1, &iod,
+			   entry->ae_sgl);
+	if (rc)
+		D_ERROR("vos_obj_fetch failed: "DF_RC"\n", DP_RC(rc));
+	entry->ae_sgl->sg_nr = AGG_IOV_CNT;
+out:
+	return rc;
+}
+
+/* Encodes a full stripe. Called when replicas form a full stripe.
+ */
+static void
+agg_encode_full_stripe(void *arg)
+{
+	struct ec_agg_entry	*entry = (struct ec_agg_entry *)arg;
+	struct obj_ec_codec	*codec;
+	unsigned int		 len = entry->ae_oca->u.ec.e_len;
+	unsigned int		 k = entry->ae_oca->u.ec.e_k;
+	unsigned int		 p = entry->ae_oca->u.ec.e_p;
+	unsigned int		 cell_bytes = len * entry->ae_rsize;
+	unsigned char		*data[k];
+	unsigned char		*parity_bufs[p];
+	unsigned char		*buf;
+	int			 i;
+
+	buf = entry->ae_sgl->sg_iovs[AGG_IOV_DATA].iov_buf;
+	for (i = 0; i < k; i++)
+		data[i] = &buf[i*cell_bytes];
+
+	buf = entry->ae_sgl->sg_iovs[AGG_IOV_PARITY].iov_buf;
+	for (i = 0; i < p; i++)
+		parity_bufs[i] = &buf[i*cell_bytes];
+
+	codec = obj_ec_codec_get(daos_obj_id2class(entry->ae_oid.id_pub));
+	ec_encode_data(cell_bytes, k, p, codec->ec_gftbls, data, parity_bufs);
+}
+
+/* Driver function for full_stripe encode. Fetches the data and then invokes
+ * second function to encode the parity.
+ */
+static int
+agg_encode_local_parity(struct ec_agg_entry *entry)
+{
+	int rc = 0;
+
+	rc = agg_fetch_data_stripe(entry);
+	if (rc)
+		goto out;
+	agg_encode_full_stripe(entry);
+
+out:
+	return rc;
+}
+
+/* True if all extents within the stripe are at a higher epoch than
+ * the parity for the stripe.
+ */
+static bool
+agg_data_is_newer(struct ec_agg_entry *entry)
+{
+	struct ec_agg_extent	*agg_extent;
+
+	d_list_for_each_entry(agg_extent, &entry->ae_cur_stripe.as_dextents,
+			      ae_link) {
+		if (agg_extent->ae_epoch < entry->ae_par_extent.ape_epoch)
+			return false;
+	}
+	return true;
+}
+
+/* Determines if the replicas present for the current stripe of object entry
+ * constitute a full stripe. If parity exists for the the stripe, the replicas
+ * making up the full stripe must be a later epoch that the parity.
+ */
+static bool
+agg_stripe_is_filled(struct ec_agg_entry *entry, bool has_parity)
+{
+	bool	is_filled, rc = false;
+
+
+	is_filled = entry->ae_cur_stripe.as_stripe_fill ==
```

Comment:
Does it assume all replica extents within same stripe are contiguous?

Reply:
Is filled is true if the sum of the sizes the individual extents for the portion of them that are within the stripe add to the stripe size.  For the this to be true, the extents must be contiguous.  So this function doesn't assume they're contiguous, but it verifies it.